data/src/main/java/org/apache/iceberg/data/GenericTaskWriter.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data;
+
+import java.io.IOException;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.io.BaseTaskWriter;
+import org.apache.iceberg.io.FileAppenderFactory;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.io.PartitionedFanoutWriter;
+import org.apache.iceberg.io.UnpartitionedWriter;
+
+public class GenericTaskWriter<T extends StructLike> extends BaseTaskWriter<T> {
```

Comment:
I understand this wraps the writer depending on if it's partitioned or not but since this is public just want to make sure it's really needed. Going through the code it seems like every engine (Spark/Flink etc) implements their own writer and specifies their own partitioning. But effectively they all follow the same principle which is implement PartitionWriter<EnginesRowRepresentation> which applies the partition function on the row representation.

The logic across the engine also looks duplicated with the same if(partition){use the partitioned writer) else {unpartitioned writer} so I think it's nice this PR has a generic writer which could be leverage. So this change makes sense to me. Maybe for partitioned writes we also allow passing in a custom partition function but we don't need to add that until we see a need.

Reply:
Thanks for review. @amogh-jahagirdar About why `GenericTaskWriter` is really needed.  Discussion above could explain some reasons.
IMO: If users need to write data into partitioned table,  they maybe need to understand `PartitionedFanoutWriter` and `InternalGenericRow`,  because genericRecord  object write directly will cause some partitioned fields cast error. About #6510 I had run into this problem, and it really confused me.  `GenericTaskWriter` reduces the difficulty for JVM applications to write to the iceberg table for users.
I have the same idea about custom partition function, add it when needed. 

Regarding the above code comment, I have solved it.